modal/_telemetry.py

```diff
@@ -0,0 +1,137 @@
+# Copyright Modal Labs 2024
+
+import importlib
+import importlib.abc
+import json
+import os
+import queue
+import socket
+import sys
+import threading
+import time
+import typing
+import uuid
+from struct import pack
+
+from typing_extensions import Self
+
+from .config import logger
+
+MODULE_LOAD_START = "module_load_start"
+MODULE_LOAD_END = "module_load_end"
+
+MESSAGE_HEADER_FORMAT = "<I"
+MESSAGE_HEADER_LEN = 4
+
+
+class ImportInterceptor(importlib.abc.Loader):
+    loading: typing.Set[str]
+    tracing_socket: socket.socket
+    events: queue.Queue
+
+    @classmethod
+    def connect(cls, socket_filename: str) -> Self:
```

Comment:
FYI Another way of returning the self type, without having to use typing_extensions for old pythons, is to string quote the type - `-> "ImportInterceptor"` in this case. Also works for forward references

Reply:
Right, I thought this felt cleaner but maybe that's a rustism.

Reply:
done